### ISSUE TITLE
Determine why OrdinalKeys aren't always respected in our studies

### ISSUE THREAD

**Issue by guidezpl** [material-components/material-components-flutter-gallery#390]
OrdinalKeys for destinations are sometimes respected, sometimes not. This leads to unexpected screenreader experience in which certain items are read out in row order, and other in column order.
## Acceptance Criteria
* Screenreader experience for Crane destinations is consistent and expected (ideally, start to end, top to bottom)

**Comment by perclasson** [comment]
I now also have the same issue in Shrine after working on:
https://github.com/material-components/material-components-flutter-gallery/issues/486

**Comment by perclasson** [comment]
Hi @goderbauer, do you have an idea for why the `OrdinalKey` is not always respected within an app?

For a smaller example, see the change I did [here](https://github.com/flutter/samples/pull/295/files). It adds OrdinalKey's to the shopping cart within the material study Shrine in the new Flutter Gallery. Even though I have added these OrdinalKey's it seems as if they're not always respected and each time I run the app with VoiceOver it can do different things (sometimes it starts at the top of the screen, sometimes at the button Clear Cart).

**Comment by goderbauer** [comment]
Can you post what the semantics tree looks like?

**Comment by goderbauer** [comment]
And is this a problem on iOS and Android or only iOS?

**Comment by perclasson** [comment]
@goderbauer The problem only seems to be for iOS, on Android it works fine. I'm running Flutter in beta.

The focus should be on the top of the screen when the cart is open:
![RPReplay_Final1581686054](https://user-images.githubusercontent.com/1770678/74534733-b01ff700-4f34-11ea-8024-f5adb17a8647.gif)

Here is the semantics tree:
```dart
flutter: SemanticsNode#0
flutter:  │ Rect.fromLTRB(0.0, 0.0, 1125.0, 2436.0)
flutter:  │
flutter:  └─SemanticsNode#1
flutter:    │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0) scaled by 3.0x
flutter:    │ textDirection: ltr
flutter:    │
flutter:    └─SemanticsNode#2
flutter:      │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:      │ flags: scopesRoute
flutter:      │
flutter:      ├─SemanticsNode#36
flutter:      │   Rect.fromLTRB(16.0, 748.0, 121.0, 796.0)
flutter:      │   flags: isButton
flutter:      │   label: "Back to Gallery"
flutter:      │   textDirection: ltr
flutter:      │   sortKey: OrdinalSortKey#c2925(order: 0.0)
flutter:      │
flutter:      └─SemanticsNode#3
flutter:        │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:        │ textDirection: ltr
flutter:        │ sortKey: OrdinalSortKey#7d52e(order: 1.0)
flutter:        │
flutter:        └─SemanticsNode#4
flutter:          │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:          │ flags: scopesRoute
flutter:          │
flutter:          ├─SemanticsNode#35
flutter:          │ │ Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:          │ │ sortKey: OrdinalSortKey#c512d(name: "home", order: 0.0)
flutter:          │ │ thickness: 4.0
flutter:          │ │
flutter:          │ ├─SemanticsNode#67
flutter:          │ │ │ Rect.fromLTRB(0.0, 44.0, 375.0, 778.0)
flutter:          │ │ │
flutter:          │ │ └─SemanticsNode#80
flutter:          │ │   │ Rect.fromLTRB(0.0, 0.0, 375.0, 734.0)
flutter:          │ │   │ flags: hasImplicitScrolling
flutter:          │ │   │ scrollChildren: 5
flutter:          │ │   │ scrollIndex: 0
flutter:          │ │   │ scrollExtentMin: 0.0
flutter:          │ │   │ scrollPosition: 0.0
flutter:          │ │   │ scrollExtentMax: 0.0
flutter:          │ │   │ elevation: 4.0
flutter:          │ │   │
flutter:          │ │   ├─SemanticsNode#68
flutter:          │ │   │   Rect.fromLTRB(0.0, 0.0, 375.0, 48.0)
flutter:          │ │   │   tags: RenderViewport.twoPane
flutter:          │ │   │   actions: tap
flutter:          │ │   │   flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          │ │   │   label:
flutter:          │ │   │     "Close cart
flutter:          │ │   │     CART
flutter:          │ │   │     1 ITEM"
flutter:          │ │   │   textDirection: ltr
flutter:          │ │   │   sortKey: OrdinalSortKey#83aef(name: "shopping_cart", order: 0.0)
flutter:          │ │   │
flutter:          │ │   ├─SemanticsNode#69
flutter:          │ │   │   Rect.fromLTRB(0.0, 48.0, 375.0, 64.0)
flutter:          │ │   │   tags: RenderViewport.twoPane
flutter:          │ │   │
flutter:          │ │   ├─SemanticsNode#70
flutter:          │ │   │ │ Rect.fromLTRB(0.0, 64.0, 375.0, 181.0)
flutter:          │ │   │ │ tags: RenderViewport.twoPane
flutter:          │ │   │ │ sortKey: OrdinalSortKey#12ffa(name: "shopping_cart", order: 1.0)
flutter:          │ │   │ │
flutter:          │ │   │ ├─SemanticsNode#71
flutter:          │ │   │ │   Rect.fromLTRB(0.0, 0.0, 60.0, 48.0)
flutter:          │ │   │ │   flags: isButton
flutter:          │ │   │ │   label: "Remove Stella sunglasses"
flutter:          │ │   │ │   textDirection: ltr
flutter:          │ │   │ │
flutter:          │ │   │ └─SemanticsNode#72
flutter:          │ │   │   │ merge boundary ⛔️
flutter:          │ │   │   │ Rect.fromLTRB(151.0, 0.0, 359.0, 36.0)
flutter:          │ │   │   │ label: "Stella sunglasses"
flutter:          │ │   │   │ textDirection: ltr
flutter:          │ │   │   │
flutter:          │ │   │   └─SemanticsNode#73
flutter:          │ │   │       merged up ⬆️
flutter:          │ │   │       merge boundary ⛔️
flutter:          │ │   │       Rect.fromLTRB(0.0, 0.0, 208.0, 17.0)
flutter:          │ │   │       label:
flutter:          │ │   │         "Quantity: 1
flutter:          │ │   │         x $58"
flutter:          │ │   │       textDirection: ltr
flutter:          │ │   │
flutter:          │ │   ├─SemanticsNode#74
flutter:          │ │   │ │ Rect.fromLTRB(0.0, 181.0, 375.0, 296.0)
flutter:          │ │   │ │ tags: RenderViewport.twoPane
flutter:          │ │   │ │ sortKey: OrdinalSortKey#4a233(name: "shopping_cart", order: 2.0)
flutter:          │ │   │ │
flutter:          │ │   │ ├─SemanticsNode#75
flutter:          │ │   │ │   merge boundary ⛔️
flutter:          │ │   │ │   Rect.fromLTRB(60.0, 0.0, 359.0, 40.0)
flutter:          │ │   │ │   label:
flutter:          │ │   │ │     "TOTAL
flutter:          │ │   │ │     $68.48"
flutter:          │ │   │ │   textDirection: ltr
flutter:          │ │   │ │
flutter:          │ │   │ ├─SemanticsNode#76
flutter:          │ │   │ │   merge boundary ⛔️
flutter:          │ │   │ │   Rect.fromLTRB(60.0, 56.0, 359.0, 73.0)
flutter:          │ │   │ │   label:
flutter:          │ │   │ │     "Subtotal:
flutter:          │ │   │ │     $58.00"
flutter:          │ │   │ │   textDirection: ltr
flutter:          │ │   │ │
flutter:          │ │   │ ├─SemanticsNode#77
flutter:          │ │   │ │   merge boundary ⛔️
flutter:          │ │   │ │   Rect.fromLTRB(60.0, 77.0, 359.0, 94.0)
flutter:          │ │   │ │   label:
flutter:          │ │   │ │     "Shipping:
flutter:          │ │   │ │     $7.00"
flutter:          │ │   │ │   textDirection: ltr
flutter:          │ │   │ │
flutter:          │ │   │ └─SemanticsNode#78
flutter:          │ │   │     merge boundary ⛔️
flutter:          │ │   │     Rect.fromLTRB(60.0, 98.0, 359.0, 115.0)
flutter:          │ │   │     label:
flutter:          │ │   │       "Tax:
flutter:          │ │   │       $3.48"
flutter:          │ │   │     textDirection: ltr
flutter:          │ │   │
flutter:          │ │   └─SemanticsNode#79
flutter:          │ │       Rect.fromLTRB(0.0, 296.0, 375.0, 396.0)
flutter:          │ │       tags: RenderViewport.twoPane
flutter:          │ │
flutter:          │ └─SemanticsNode#81
flutter:          │     Rect.fromLTRB(16.0, 714.0, 359.0, 762.0)
flutter:          │     actions: tap
flutter:          │     flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          │     label: "CLEAR CART"
flutter:          │     textDirection: ltr
flutter:          │     elevation: 4.0
flutter:          │     thickness: 2.0
flutter:          │
flutter:          └─SemanticsNode#5
flutter:              Rect.fromLTRB(0.0, 0.0, 375.0, 812.0)
flutter:              sortKey: OrdinalSortKey#963cb(name: "home", order: 1.0)
flutter:
```

`flutter doctor -v`:
```dart
[✓] Flutter (Channel beta, v1.14.6, on Mac OS X 10.15.2 19C57, locale en-DE)
    • Flutter version 1.14.6 at /Users/perc/development/flutter
    • Framework revision fabeb2a16f (2 weeks ago), 2020-01-28 07:56:51 -0800
    • Engine revision c4229bfbba
    • Dart version 2.8.0 (build 2.8.0-dev.5.0 fc3af737c7)


[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/perc/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-stable, build-tools 29.0.2
    • Java binary at: /Applications/Android Studio with Blaze.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.3.1, Build version 11C505
    • CocoaPods version 1.8.4

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 3.6)
    • Android Studio at /Applications/Android Studio with Blaze.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b4-5784211)

[✓] Android Studio (version 3.5)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 43.0.1
    • Dart plugin version 191.8593
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[!] VS Code (version 1.41.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    ✗ Flutter extension not installed; install from
      https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (5 available)
    • Pixel 4           • 9B011FFAZ00072                       • android-arm64  • Android 10 (API 29)
    • iPhone            • 00008030-0016518E34D2802E            • ios            • iOS 13.2.3
    • Chrome            • chrome                               • web-javascript • Google Chrome 80.0.3987.106
    • Web Server        • web-server                           • web-javascript • Flutter Tools
```

**Comment by goderbauer** [comment]
Semantics Tree looks correct. Unfortunately, a11y focus order is hard to control on iOS. They don't offer an API for it. 

In theory, iOS should select the first a11y node when the `UIAccessibilityScreenChangedNotification` is triggered (which Flutter triggers when the topmost route changes). I can see two potential failures here:

* Flutter doesn't recognize opening the shopping cart as a new route. You can test this by throwing a print on [this line](https://github.com/flutter/engine/blob/cd77e784309cfe91ec9d207fb55631d3392964e5/shell/platform/darwin/ios/framework/Source/accessibility_bridge.mm#L821) here and see if it triggers when you open the shopping card. If it doesn't trigger you'll have to add a Semantics widget with `scopesRoute` and maybe `namesRoute` set to true (@jonahwilliams may be able to help with fixing that).
* Flutter recognizes the new route, but we dispatch the `UIAccessibilityScreenChangedNotification` too early at a time where the new route hasn't created all its SemanticsNodes yet. This one would probably require some kind of a fix in the Flutter engine, but it will be hard.

**Comment by willlarche** [comment]
Closing as intended behavior.
